### PR TITLE
Options for ListItemPicker (substringof  search) and ListItemAttachments (open in new window)

### LINKED ISSUE #582: ListItemAttachments fails in Microsoft Teams Tab SPFx applications
#### Category
[ ] Enhancement

[X ] Bug

[ ] Question

#### Version
Please specify what version of the library you are using: [1.18.0]

#### Expected / Desired Behavior / Question
ListItemAttachments control should work on a Teams Tab. The window.location change in the onClickHref handler won't work in Microsoft Teams due to IFRAME security.

#### Observed Behavior
Clicking on an attachment icon in view or edit mode from a Tab in Microsoft Teams causes the Tab to crash, and the navigation attempt getting blocked.

#### Steps to Reproduce
Build an SPFx Teams Tab app that uses the ListItemAttachments control. Add an attachment. Then try opening the attachment.

Joel Jeffery
[w:joelblogs.co.uk](https://joelblogs.co.uk)
[t:@joelblogs](https://twitter.com/joelblogs)

### LINKED ISSUE #583: ListItemPicker could use a substring search instead of just startswith
#### Category
[X] Enhancement

[ ] Bug

[ ] Question

#### Version

Please specify what version of the library you are using: [1.18.0]

#### Expected / Desired Behavior / Question
I'd like the option of finding List Items in picker as a substring search rather than just starts-with.

#### Observed Behavior
Users need to know exactly how an item in the target list _starts_ in order to find it. This is only feasible for some use cases.

#### Steps to Reproduce
Create a List with many items with very different titles. Maybe all the city names in your country. Or dog breeds. Or cheeses. I'm pretty sure there's some CSVs of those somewhere 😉. Create a List with a lookup column into the first list. Build a test app/web part or something to use the ListItemPicker control on the lookup field.

Try and find a city/dog/cheese :)

Now, wouldn't that be better with a substring search?

You betcha.

Joel Jeffery
[w:joelblogs.co.uk](https://joelblogs.co.uk)
[t:@joelblogs](https://twitter.com/joelblogs)

### LINKED ISSUE #585: Map Control doesn't return display name or address details of locations
#### Category
[X] Enhancement

[ ] Bug

[ ] Question

#### Version

Please specify what version of the library you are using: [1..0.18]

#### Expected / Desired Behavior / Question
The Map control uses OpenStreetMap.org's nominatim web service to perform searches and get location data. This returns a display name and can return address details.

SharePoint "Location" fields could use this data - albeit in a slightly different format.

Let's return this data along with the coordinates.

#### Observed Behavior
Only coordinates are returned by the Map.

#### Steps to Reproduce
Build an SPFx web part, hook up the Map control. Try and use the returned data to make a suitable value for a SharePoint Location field. Observe you could really use more data if it's there.

Joel Jeffery
[w:joelblogs.co.uk](https://joelblogs.co.uk)
[t:@joelblogs](https://twitter.com/joelblogs)

### LINKED ISSUE #586: Support for Bing Maps - draggable and static
#### Category
[X] Enhancement

[ ] Bug

[ ] Question

#### Version

Please specify what version of the library you are using: [1.18.0]

#### Expected / Desired Behavior / Question
Sometimes OpenStreetMaps fails to zoom. Right clicking the map and reloading the IFRAME in the browser makes the problem go away.

As an alternative, it would be nice to support Bing Maps to display the results of the nominatim web service call. 

#### Observed Behavior
Map sometimes fails to zoom, showing whole globe when coordinates and zoom level are passed in as props.

#### Steps to Reproduce
Pass in coordinates as props; pass in a zoom level. Observe occasional failure to zoom.

[reference](https://help.openstreetmap.org/questions/20335/embedded-html-displays-zoomed-out)

### PR DESCRIPTION
**Pull request by joeljeffery**
| Q               | A
| --------------- | ---
| Bug fix?        | [X ]
| New feature?    | [X ]
| New sample?      | [ ]
| Related issues?  | fixes #582, fixes #583, fixes #585, fixes #586 

#### What's in this Pull Request?

**Four changes:**
_ListItemPicker - option to search by substring_
- added substringSearch?:boolean prop
- added substringSearch:boolean = false parameter to SPService.getListItems 
- added substringOf as optional search strategy to SPService.getListItems 

_ListItemAttachments - option to open attachments in new window to support use from Microsoft Teams_
- added openAttachmentsInNewWindow?: boolean prop
- added onClick handler window.open(`${file.ServerRelativeUrl}?web=1`, "_blank")

_Map - return display name and address details along with coordinates_
- added address: any to LocationInfo interface
- added address and displayname fields to ICoordinates interface
- added &addressdetails=1 parameter to nominatim call
- added returned display_name and address from nominatim data as displayname and address fields on coordinates object returned from map
- added type to coordinates object passed to onUpdateCoordinates() in _getCoordinates()
- added support for zoom levels beyond 15

_Map - added rudimentary support for Bing maps_
- added alternative _getBingMapUrl method to return suitable draggable or static bing.com/maps/embed link (no API key required)
- added pushpin indicator icon for Bing static (non-draggable) map  to centre of map

Joel Jeffery
[w:joelblogs.co.uk](https://joelblogs.co.uk)
[t:@joelblogs](https://twitter.com/joelblogs)

### CODE REVIEW

File: src/controls/map/Map.tsx [hunk 1]
```diff
@@ -35,29 +33,34 @@ export class Map extends React.Component<IMapProps, IMapState> {
   /**
    * componentWillMount lifecycle hook
    */
-  public componentWillMount(): void {
-    this.setState({
-      coordinates: this.props.coordinates
-    });
-  }
+  // 20200614 - legacy react lifecycle event; use constructor instead
```
Comment (AJIXuMuK) [inline src/controls/map/Map.tsx:36]: We're still using React v 15 in this library, so it's not legacy for this version.
Anyway, we can remove this one.

**Comment by AJIXuMuK** [comment]
Hi @joeljeffery!
Thank you for all the changes!

I've added some comments to check.

And it would be also great if you merge with `dev` first as there are conflicts right now.

File: src/controls/map/Map.tsx [hunk 2]
```diff
@@ -35,29 +33,34 @@ export class Map extends React.Component<IMapProps, IMapState> {
...
   /**
   * Get the dif value based on zoom supplied (dif is for calculating the 4 corners of the map)
   */
   private _getDif(): number {
-    let zoom: number = this.props.zoom >= 0 ? this.props.zoom : 10;
-    return 0.0025 + (0.005 * (15 - zoom));
+    const { zoom } = this.props;
+    // 20200614 - JJ - support zoom levels beyond 15
+    const newZoom: number = zoom >= 0 ? zoom % 16 : 10;
+    const multiplier = Math.floor(zoom / 16) + 1;
```
Comment (AJIXuMuK) [inline src/controls/map/Map.tsx:62]: if `zoom` is `undefined` - this line will crash the map with `NaN` coordinates

**Comment by AJIXuMuK** [comment]
@joeljeffery - thank you for the changes. I have resolved the issues and conflicts and merged the PR.

**Comment by joeljeffery** [comment]
>And it would be also great if you merge with dev first as there are conflicts right now.

Thanks @AJIXuMuK for rolling my changes in!

I'll be sure to use the dev branch next time! 😊

**Pull request merged**